File: src/Desktop/Networking/ApiService.cs
using Noteapp.Desktop.Data;
using Noteapp.Desktop.Dtos;
using Noteapp.Desktop.Extensions;
using Noteapp.Desktop.Models;
using Noteapp.Desktop.Security;
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Threading.Tasks;

namespace Noteapp.Desktop.Networking
{
    public class ApiService
    {
        private readonly HttpClient _httpClient;

        public ApiService(HttpClient httpClient)
        {
            _httpClient = httpClient ?? throw new ArgumentNullException(nameof(httpClient));
        }

        public async Task<IEnumerable<Note>> GetNotes()
        {
            var request = new HttpRequestMessage(HttpMethod.Get, "notes");
            var response = await SendRequest(request);
            return await ReadNotesFromResponse(response);
        }

        public async Task<Note> CreateNote(Note note)
        {
            var request = new HttpRequestMessage(HttpMethod.Post, "notes");
            return await SendRequestAndReadNoteFromResponse(request, note);
        }

        public async Task<bool> BulkCreateNotes(IEnumerable<Note> notes)
        {
            var noteRequests = new List<NoteRequest>();

            foreach (var note in notes)
            {
                var noteRequest = new NoteRequest(note);
                noteRequest.Text = await Protector.TryEncrypt(noteRequest.Text);
                noteRequests.Add(noteRequest);
            }

            var request = new HttpRequestMessage(HttpMethod.Post, "notes/bulk");
            request.Content = JsonContent.Create(noteRequests);
            return (await SendRequest(request)).IsSuccess;
        }

        public async Task<Note> UpdateNote(Note note)
        {
            var request = new HttpRequestMessage(HttpMethod.Put, $"notes/{note.Id}");
            return await SendRequestAndReadNoteFromResponse(request, note);
        }

        public async Task<bool> DeleteNote(int noteId)
        {
            var request = new HttpRequestMessage(HttpMethod.Delete, $"notes/{noteId}");
            return (await SendRequest(request)).IsSuccess;
        }

        public async Task<IEnumerable<NoteSnapshot>> GetAllSnapshots(int noteId)
        {
            var request = new HttpRequestMessage(HttpMethod.Get, $"notes/{noteId}/snapshots");
            var response = await SendRequest(request);
            return await ReadSnapshotsFromResponse(response);
        }

        public async Task<RegisterResult> Register(string email, string password)
        {
            var request = new HttpRequestMessage(HttpMethod.Post, "account/register");
            request.Content = JsonContent.Create(new { email, password });
            var response = await SendRequest(request, false);

            var result = new RegisterResult();
            result.ErrorMessage = response.ErrorMessage;
            return result;
        }

        public async Task<LoginResult> Login(string email, string password)
        {
            var request = new HttpRequestMessage(HttpMethod.Post, "account/token");
            request.Content = JsonContent.Create(new { email, password });
            var response = await SendRequest(request, false);

            var result = new LoginResult();
            result.ErrorMessage = response.ErrorMessage;
            result.UserInfoResponse = response.IsSuccess ?
                response.Content.FromJson<UserInfoResponse>() : null;
            return result;
        }

        public async Task<bool> DeleteAccount()
        {
            var request = new HttpRequestMessage(HttpMethod.Delete, "account/delete");
            return (await SendRequest(request)).IsSuccess;
        }

        private async Task<ApiResponse> SendRequest(HttpRequestMessage request, bool authorized = true)
        {
            if (authorized) AddAuthorizationHeader(request);

            HttpResponseMessage response;
            var apiResponse = new ApiResponse();
            try
            {
                response = await _httpClient.SendAsync(request);
            }
            catch (HttpRequestException)
            {
                apiResponse.ErrorMessage = "Failed to connect to the server";
                return apiResponse;
            }

            var content = await response.Content.ReadAsStringAsync();
            if (response.IsSuccessStatusCode)
            {
                apiResponse.Content = content;
            }
            else
            {
                apiResponse.ErrorMessage = !string.IsNullOrWhiteSpace(content)
                    ? content.FromJson<ErrorResponse>().ErrorMessage : "ERROR";
            }

            return apiResponse;
        }

        private void AddAuthorizationHeader(HttpRequestMessage request)
        {
            var accessToken = AppData.UserInfo.AccessToken;
            if (!string.IsNullOrWhiteSpace(accessToken))
            {
                request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);
            }
        }

        private async Task<Note> SendRequestAndReadNoteFromResponse(HttpRequestMessage request, Note note)
        {
            var noteRequest = new NoteRequest(note);
            noteRequest.Text = await Protector.TryEncrypt(noteRequest.Text);
            request.Content = JsonContent.Create(noteRequest);
            var response = await SendRequest(request);

            if (!response.IsSuccess) return null;
            var updatedNote = response.Content.FromJson<Note>();
            updatedNote.Text = await Protector.TryDecrypt(updatedNote.Text);
            return updatedNote;
        }

        private async Task<IEnumerable<Note>> ReadNotesFromResponse(ApiResponse response)
        {
            if (!response.IsSuccess) return null;
            var notes = response.Content.FromJson<IEnumerable<Note>>();
            foreach (var note in notes)
            {
                note.Text = await Protector.TryDecrypt(note.Text);
            }
            return notes;
        }

        private async Task<IEnumerable<NoteSnapshot>> ReadSnapshotsFromResponse(ApiResponse response)
        {
            if (!response.IsSuccess) return null;
            var snapshots = response.Content.FromJson<IEnumerable<NoteSnapshot>>();
            foreach (var snapshot in snapshots)
            {
                snapshot.Text = await Protector.TryDecrypt(snapshot.Text);
            }
            return snapshots;
        }
    }
}

File: src/Web/wwwroot/js/ApiService.js
import AppData from "./AppData.js";
import Config from "./Config.js";

let ApiService = {
    getNotes,
    createNote,
    updateNote,
    deleteNote,
    getAllSnapshots,
    login,
    register,
    deleteAccount
}

async function getNotes() {
    let response = await sendRequest("notes", "GET");
    if (!response.isSuccess()) return null;

    let notes = response.content;
    notes.forEach(note => {
        note.synchronized = true;
        note.updatedLocal = note.updated;
    });
    return notes;
}

async function createNote(note) {
    let headers = { "Content-Type": "application/json" };
    let body = JSON.stringify(new NoteRequest(note));
    let response = await sendRequest("notes", "POST", headers, body);
    return await getNoteFromResponse(response);
}

async function updateNote(note) {
    let headers = { "Content-Type": "application/json" };
    let body = JSON.stringify(new NoteRequest(note));
    let response = await sendRequest(`notes/${note.id}`, "PUT", headers, body);
    return await getNoteFromResponse(response);
}

async function deleteNote(noteId) {
    return (await sendRequest(`notes/${noteId}`, "DELETE")).isSuccess();
}

async function getAllSnapshots(noteId) {
    let response = await sendRequest(`notes/${noteId}/snapshots`);
    return response.isSuccess() ? response.content : null;
}

async function register(email, password) {
    let headers = { "Content-Type": "application/json" };
    let body = JSON.stringify({ email, password });
    return await sendRequest("account/register", "POST", headers, body, false);
}

async function login(email, password) {
    let headers = { "Content-Type": "application/json" };
    let body = JSON.stringify({ email, password });
    return await sendRequest("account/token", "POST", headers, body, false);
}

async function deleteAccount() {
    return (await sendRequest("account/delete", "DELETE")).isSuccess();
}

async function sendRequest(url, method, headers = {}, body = null, authorized = true) {
    if (authorized) addAuthorizationHeader(headers);

    let response;
    let apiResponse = new ApiResponse();
    try {
        response = await fetch(`${Config.API_BASE_URL}${url}`, { method, headers, body });
    }
    catch {
        apiResponse.errorMessage = 'Failed to connect to the server';
        return apiResponse;
    }

    let content = await response.text();
    if (response.ok) {
        apiResponse.content = content.length > 0 ? JSON.parse(content) : null;
    }
    else {
        apiResponse.errorMessage = content.length > 0 ?
            JSON.parse(content).errorMessage : "ERROR";
    }
    return apiResponse;
}

function addAuthorizationHeader(headers) {
    let userInfo = AppData.getUserInfo();
    if (userInfo.accessToken != null) {
        headers['Authorization'] = `Bearer ${userInfo.accessToken}`;
    }
}

async function getNoteFromResponse(response) {
    if (!response.isSuccess()) return null;

    let note = response.content;
    note.synchronized = true;
    note.updatedLocal = note.updated;
    return note;
}

class ApiResponse {
    content = null;
    errorMessage = null;
    constructor() {}
    isSuccess() {
        return this.errorMessage == null || this.errorMessage == '';
    }
}

function NoteRequest(note) {
    this.text = note.text;
    this.locked = note.locked;
    this.archived = note.archived;
    this.pinned = note.pinned;
    this.published = note.published;
}

export default ApiService;